Object methods, "this" (#343)
javascript-tutorial/es.javascript.info: 1-js/04-object-basics/04-object-methods/article.md

> # Métodos de objeto, "this"

Reviewer: ```suggestion
# Métodos del objeto, "this"
```

};
```

> Y, en el mundo real, un usuario puede *actuar*: seleccionar algo del carrito de compras, hacer login, logout, etc.

Author: @maksumi  Depende de la hora en que escribo
sabía que cualquiera lo iba a ignorar y vos lo ibas a acotar...
pero no recuerdo la respuesta preparada  :(

igual dejarlas sería de caprichoso, molestan, es MUCHO mejor sin ellas.

*!*
user.sayHi = function() {
  alert("¡Hola!");
};
*/!*

user.sayHi(); // ¡Hola!
```

Aquí simplemente usamos una expresión de función para crear la función y asignarla a la propiedad `user.sayHi` del objeto.

Entonces la llamamos. ¡El usuario ahora puede hablar!

Una función que es la propiedad de un objeto es denominado su *método*.

Así, aquí tenemos un método `sayHi` del objeto `user`.

> Por supuesto, podríamos usar una función pre-declarada como un método, como esto:

Reviewer: ```suggestion
Por supuesto, podríamos usar una función pre-declarada como un método, parecido a esto:
```

```js run
let user = {
  // ...
};

*!*
// primero, declara
function sayHi() {
  alert("¡Hola!");
};

// entonces la agrega como un método
user.sayHi = sayHi;
*/!*

user.sayHi(); // ¡Hola!
```

```smart header="Programación orientada a objetos"

> Cuando escribimos nuestro código usando objetos que representan entidades, eso es llamado [Programación orientada a objetos](https://es.wikipedia.org/wiki/Programaci%C3%B3n_orientada_a_objetos), abreviado: "POO".

Author: @maksumi  
En realidad no.  Lo acepto por cómo se ve y nadie se va a quejar.
En ingles sí se ponen todos los sustantivos en maúsculas, en español no.
Programacion lo puse en mayusc porque quedaba lindo, destacado como link, tampoco segui las reglas.

> POO (OOP sus siglas en inglés) es una cosa grande, un ciencia interesante en sí misma. ¿Cómo elegir las entidades correctas? ¿Cómo organizar la interacción entre ellas? Eso es arquitectura, y hay muy buenos libros del tópico como "Patrones de diseño: Elementos de software orientado a objetos reutilizable" de E. Gamma, R. Helm, R. Johnson, J. Vissides o "Análisis y Diseño Orientado a Objetos" de G. Booch, y otros.

Author: @maksumi 
Yo sí, corté y pegué.
Me tomé el trabajo de buscarlos porque me parecio muy importante, 
es buena recomendacion:  Si LE HACE CASO, los querrá en español, y existen.
Podria agregar un "Espero que los puedas conseguir en tu país"

> POO (OOP sus siglas en inglés) es una cosa grande, un ciencia interesante en sí misma. ¿Cómo elegir las entidades correctas? ¿Cómo organizar la interacción entre ellas? Eso es arquitectura, y hay muy buenos libros del tópico como "Patrones de diseño: Elementos de software orientado a objetos reutilizable" de E. Gamma, R. Helm, R. Johnson, J. Vissides o "Análisis y Diseño Orientado a Objetos" de G. Booch, y otros.

Reviewer 2: That's ok! Disculpa las deficiencias de mi review, pero ya tengo internet, no volverá a pasar jejeje

```
### Formas abreviadas para los métodos

Existe una sintaxis más corta para los métodos en objetos literales:

```js
// estos objetos hacen lo mismo

user = {
  sayHi: function() {
    alert("Hello");
  }
};

> // abreviado se ve mejor, ¿verdad?

Reviewer: ```suggestion
// la forma abreviada se ve mejor, ¿verdad?
```

user = {
*!*
  sayHi() { // igual que "sayHi: function()"
*/!*
    alert("Hello");
  }
};
```

Como se demostró, podemos omitir `"function"` y simplemente escribir `sayHi()`.

A decir verdad, las notaciones no son completamente idénticas. Hay diferencias sutiles relacionadas a la herencia de objetos (por cubrir más adelante) que no importan ahora. En casi todos los casos la sintaxis abreviada es la preferida.

## "this" en métodos

Es común que in método de objeto necesite acceder a la información almacenada en el objeto para cumplir su tarea.

Por ejemplo, el código dentro de `user.sayHi()` puede necesitar el nombre del usuario `user`.

> **Para acceder al objeto, un método usa la palabra clave `this`.**

Reviewer: ```suggestion
**Para acceder al objeto un método puede usar la palabra clave `this`.**
```

> **Para acceder al objeto un método puede usar la palabra clave `this`.**

Author: !!!
es que tengo que limpiar el teclado...

user.sayHi(); // John
```

Aquí durante la ejecución de `user.sayHi()`, el valor de `this` será `user`.

> Técnicamente, es también posible acceder al objeto sin `this`, haciendo referencia a él por medio de la variable externa:

Reviewer 2: ```suggestion
Técnicamente, también es posible acceder al objeto sin `this`, haciendo referencia a él por medio de la variable externa:
```

  sayHi() {
*!*

>     alert(user.name); // "user" en vez de"this"

Reviewer 2: ```suggestion
    alert(user.name); // "user" en vez de "this"
```

*/!*
  }

};
```

...Pero tal código no es confiable. Si decidimos copiar `user` a otra variable, por ejemplo `admin = user` y sobrescribir `user` con otra cosa, entonces accederá al objeto incorrecto.

> Eso queda demostrado aquí:

Reviewer: ```suggestion
Eso queda demostrado en las siguientes lineas:
```

  sayHi() {
*!*
    alert( user.name ); // lleva a un error
*/!*
  }

};

let admin = user;
user = null; // sobrescribimos para hacer las cosas evidentes

> admin.sayHi(); // ¡oops! dentro de sayHi(), ¡el nombre viejo es usado! ¡Error!

Reviewer 2: ```suggestion
admin.sayHi(); // ¡oops! dentro de sayHi(), ¡el nombre viejo esta en uso! ¡Error!
```
A parte de viejo lo usan... ¡¿Por qué son así?!

}

*!*
// usa la misma función en dos objetos
user.f = sayHi;
admin.f = sayHi;
*/!*

// estos llamados tienen diferente "this"

> // "this" dentro de la función es el objeto "delante del punto"

Reviewer 2: ```suggestion
// "this" dentro de la función es el objeto "delante del punto"
```
¿Yo estoy mal? ¿Es mi escasa comprensión del inglés? ¿Por qué "before the dot" es "delante del punto" si el texto en inglés concuerda con el ejemplo en qué `this` es el objeto antes del punto?

 `this == user` y `user` está antes del punto

sayHi(); // undefined
```

En este caso `this` es `undefined` en el modo estricto. Si tratamos de acceder a `this.name`, habrá un error.

En modo no estricto el valor de `this` en tal caso será el *objeto global* (`window` en un navegador, llegaremos a ello en el capítulo [](info:global-object)). Este es un comportamiento histórico que `"use strict"` corrije.

> Usualmente tal llamado en un error de programa. Si hay `this` dentro de una función, se espera que sea llamada en un contexto de objeto.

Reviewer 2: ```suggestion
Usualmente tal llamado es un error de programa. Si hay `this` dentro de una función, se espera que sea llamada en un contexto de objeto.
```

````

```smart header="Las consecuencias de un `this` desvinculado"
Si vienes de otro lenguaje de programación, probablemente estés habituado a la idea de un "`this` vinculado", donde los método definidos en un objeto siempre tienen `this` referenciando ese objeto.

En JavaScript `this` es "libre", su valor es evaluado al momento de su llamado y no depende de dónde fue declarado el método sino de cuál es el objeto "delante del punto".

El concepto de `this` evaluado en tiempo de ejecución tiene sus pros y sus contras. Por un lado, una función puede ser reusada por diferentes objetos. Por otro, la mayor flexibilidad crea más posibilidades para equivocaciones.

Nuestra posición no es juzgar si la decisión del diseño de lenguaje es buena o mala. Vamos a entender cómo trabajar con ello, obtener beneficios y evitar problemas.
```

## Las funciones de flecha no tienen "this"

> Las funciones de flecha son especiales: ellas no tienen su "propio" `this`. Si referenciamos `this` desde tales funciones, será tomado desde el exterior "normal" de la función.

Reviewer: ```suggestion
Las funciones de flecha son especiales: ellas no tienen su "propio" `this`. Si nosotros hacemos referencia a `this` desde tales funciones, esta será tomada desde afuera función "normal".
```

Author: @danilobrinu 
Estaba mal, pero aún suena raro...  238 Creo que es por esto solo: "de la" 
desde afuera función "normal".

Reviewer: Me comi el de la jaja
